### PR TITLE
[CWS][SEC-8348] Apply rules on snapshot data

### PR DESCRIPTION
**Pull request by mftoure**
This PR aims to apply rules on snapshot data by creating fake exec events.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/security/secl/model/model.go [hunk 1]
```diff
@@ -778,8 +778,17 @@ type ProcessCacheEntry struct {
 	refCount  uint64                     `field:"-" json:"-"`
 	onRelease func(_ *ProcessCacheEntry) `field:"-" json:"-"`
 	releaseCb func()                     `field:"-" json:"-"`
+	ProcessCacheEntrySource
 }
 
+type ProcessCacheEntrySource uint64
```
Comment (safchain) [inline pkg/security/secl/model/model.go:784]: I would put this in the ProcessContext, just calling it `Source`

File: pkg/eventmonitor/eventmonitor.go [hunk 2]
```diff
@@ -155,6 +155,24 @@ func (m *EventMonitor) Start() error {
 		}
 	}
 
+	// Get the snapshotted data
+	entrycache := m.Probe.GetResolvers().ProcessResolver.GetEntryCache()
```
[the remark this answers is not in the document]
Comment (mftoure) [reply pkg/eventmonitor/eventmonitor.go:159]: At the beginning of probe.Start, the rules are not loaded yet. Is there a reason why we start the probe before starting the eventConsumers (where the rules are loaded) ?

File: pkg/security/secl/model/model.go (hunk 1 above)
Comment (mftoure) [reply pkg/security/secl/model/model.go:784]: yep

**Review by brycekahle — APPROVED** [review]

File: pkg/security/probe/probe_linux.go [hunk 3]
```diff
@@ -303,6 +306,31 @@ func (p *Probe) Start() error {
 	})
 }
 
+func (p *Probe) PlaySnapshot() {
+	p.snapshotReplayLock.Lock()
```
Comment (safchain) [inline pkg/security/probe/probe_linux.go:310]: I'm not sure a mutex here, but you need one in the GetEntryCache

File: pkg/security/secl/model/model.go [hunk 4]
```diff
@@ -782,6 +782,12 @@ type ProcessCacheEntry struct {
 	releaseCb func()                     `field:"-" json:"-"`
 }
 
+const (
+	ProcessCacheEntryFromEvent     uint64 = 1
```
Comment (safchain) [inline pkg/security/secl/model/model.go:786]: nit: you could use `iota`

File: pkg/security/secl/model/model.go [hunk 5]
```diff
@@ -864,6 +870,7 @@ func (p *ProcessContext) HasParent() bool {
 // ProcessContext holds the process context of an event
 type ProcessContext struct {
 	Process
+	Source uint64 `field:"-" json:"-"`
```
Comment (YoannGh) [inline pkg/security/secl/model/model.go:873]: I found the source information valuable when looking at generated events. Could you add this field to the `ProcessContextSerializer` as well?

Comment (mftoure) [reply pkg/security/secl/model/model.go:873]: sure !

**Review by alai97 — APPROVED** [review]
Looks good for docs!

File: pkg/security/resolvers/process/resolver.go [hunk 6]
```diff
@@ -1217,6 +1210,15 @@ func (p *Resolver) GetEntryCacheSize() float64 {
 	return float64(p.cacheSize.Load())
 }
 
+// GetEntryCache return the resolver's entrycache
+func (p *Resolver) GetEntryCache(entryToEvent func(_ *model.ProcessCacheEntry)) {
```
Comment (safchain) [inline pkg/security/resolvers/process/resolver.go:1214]: Sorry, I didn't see that this function was existing already existing : https://github.com/DataDog/datadog-agent/blob/main/pkg/security/resolvers/process/resolver.go#L1226

File: pkg/security/secl/model/model.go [hunk 7]
```diff
@@ -782,6 +782,22 @@ type ProcessCacheEntry struct {
...
+	ProcessCacheEntryFromProcFS
+)
+
+var ProcessSources = [...]string{
+	"Event",
+	"KernelMap",
+	"ProcFS",
+}
+
+func ProcSourceToString(source uint64) string {
```
Comment (safchain) [inline pkg/security/secl/model/model.go:797]: nit: ProcessSourceToString

File: pkg/security/resolvers/process/resolver.go (hunk 6 above)
Comment (mftoure) [reply pkg/security/resolvers/process/resolver.go:1214]: Wow ! I did not even see it

**Review by safchain — APPROVED** [review]

**Pull request merged**